### ISSUE TITLE
Class structure for the cards

### ISSUE THREAD

**Issue by Turtyo** [Saplings-Projects/1M_sub#14]
## Feature Description

What should be the structure of the classes for the cards to be able to have multiple effects, such as damage and poison or damage and heal for example ?

## Motivation

This issue is meant as a way to discuss our structure going forward. Right now, `CardBase` handles most of that, but as a parent class of other effect, it shouldn't have functions that are not useful to multiple children classes (for example, it shouldn't have a `drawCard` or `Heal` function in it, those should be defined in specific children classes.

## Acceptance Criteria

- `CardBase` should hold function that are useful to multiple children classes, and not specific effects
- It should be easy to add effects to a card

## Proposed Solution

We would have:
```GDscript
# CardBase.gd
extends Node

# Base class for cards
class_name CardBase

var card_effects = []

func _ready():
    # Initialize a new array for each instance
    card_effects = []

func apply_all_effects(card):
    for effect in card.card_effects:
        effect.apply_effect()
```
with:
```GDscript
# CardEffect.gd
extends Node

# Base class for card effects
class_name CardEffect

fund _ready(card):
    # to be used by child classes to add their effect to the card
    card.card_effects.append(self)

func apply_effect():
    pass
```
and the effects would extend from `CardEffect`:
```GDscript
# CardPoison.gd
extends CardEffect

class_name CardPoison

# @Override
func apply_effect():
    # do some poison damage
```

When you want to create a card, you can either use a JSON format like this:
```json
[
  {
    "name": "CardWithPoisonAndHeal",
    "description": "Apply 3 poison and heal 3",
    "effects": [
      {
        "type": "CardPoison",
        "parameters": {
          "duration": 3,
          "damage_per_turn": 3
        }
      },
      {
        "type": "CardHeal",
        "parameters": {
          "heal_amount": 3
        }
      }
    ]
  }
]
```

And we would have some code to display the card / setting the effects in the background by reading the file
But i would like to not have that be the structure, because JSON files are not very nice to work with. For example i could say 3 poison in the description but put 4 by mistake in the duration.

Ideally, it would be good to have something that's directly editable. For example we could have something in the text file of the card that reads:
"Apply PoisonEffect(3) and HealEffect(4)". That would setup a card with 3 poison and 4 heal, but when displayed it would read as "Apply 3 poison stacks and heals 4". This could probably be done by using "to_str" and by using a parser. This probably should be addressed in another issue

## Additional Context

The discussion was started in #11

**Comment by cheesefrycook** [comment]
I think this is a pretty good idea. I still think I prefer the current implementation, but I can see why some people may prefer this one.

I think if we do this we should have hierarchy of CardEffects. Base effects would be basic actions a card can do, like dealing damage, drawing cards, applying buffs, and restoring AP. The hierarchy would look something like this:
```
CardEffect
| Effect_Damage (deals X damage when the effect is applied)
|-- Effect_DamageEqualAP (deals damage equal to AP when the effect is applied)
| Effect_DrawCards (draws X cards when the effect is applied)
|-- Effect_DrawHalfHealth (draws 1 card. Draws 1 more card if less than half health)
| Effect_RestoreAP
|-- ...
| Effect_ApplyBuff
|-- ...
```
This would prevent us having to rewrite the basic action in every type of card, and instead allow us to call a parent function that does the action. For example, in the parent `Effect_Damage`, we would have a `_deal_damage(amount: int)` function that calls `HealthComponent.deal_damage(damage_data)`. In cards that are children of Effect_Damage, we just call `_deal_damage(amount: int)`. Maybe this overcomplicates the structure, but could help with reducing error.

**Pros of this system:**
- Stays within OOP principles, as opposed to the current system. Cards will only align with what they are actually able to do. A poison card doesn't have the ability to draw cards, for example. This reduces a lot of unnecessary code in the parent `CardBase` class - instead moving that code to different CardEffects.
- Could speed up creation of cards in the long run, because we will have a library of CardEffects that we can stack together to make new cards.
- A unique CardEffect could be reused if the same effect is used for a different card. In the current system, unique effects are rewritten per card.

**Cons of this system:**
- Will increase the amount of Resources in our project. In the current system, one card needs one CardBase Resource. In this system, a card needs a CardBase Resource, and potentially multiple CardEffect Resources.
- More complicated structure. Instead of everything being derived from CardBase, cards now have multiple CardEffects that have a hierarchy. This could be confusing for new people joining the project, and more difficult to debug.
- Cards can have multiple effects that do practically the same thing. For example, "Deal 3 damage. If the enemy has a debuff, draw 1 card and restore 1 AP." This card has 3 CardEffects.
  1) Effect 1: Effect_Damage -> deal 2 damage
  2) Effect 2: derived from Effect_DrawCards -> draw 1 card if target has a debuff
  3) Effect 3: derived from Effect_RestoreAP -> restore 1 AP if the target has a debuff
  - In the current system, you would create a new child of CardBase, then do all three things at the same time.

**Comment by cheesefrycook** [comment]
Also, I don't really see the benefit of using a JSON file. The way Godot Resources work, you can create an instance of it which will give you a new file in the editor. Then with that new Resource file, you can easily edit the properties in the Godot inspector. Resources are also human readable I believe, so you can use an external editor if you want. So my thought is we have Resource instances for each card in the game.

I think the string parser is a neat idea. Could be quite tricky to implement though.

**Comment by Turtyo** [comment]
I don't really see the idea being this structure:
```
CardEffect
| Effect_Damage (deals X damage when the effect is applied)
|-- Effect_DamageEqualAP (deals damage equal to AP when the effect is applied)
| Effect_DrawCards (draws X cards when the effect is applied)
|-- Effect_DrawHalfHealth (draws 1 card. Draws 1 more card if less than half health)
| Effect_RestoreAP
|-- ...
| Effect_ApplyBuff
|-- ...
```
Why not just say it's a damage card that does 3 damage, or a draw card that draws 2 ? No need to have sub-categories like that, it will just complicate the structure.
You say that way children cards will just be able to call the action of the parent class, But the way I explained above, a card would have damage and draw effect for example, and it would just call the function of each relevant class directly.
Here, in my opinion, having more classes is the way in which you end up with more errors.

I agree for the pros, regarding the cons:
- Having everything under the same resource is not necessarily a better thing, so i don't thing it's a con to have more resources if it allows to not dump everything in the same class.
- You talk about a hierarchy, but you are the one that gave the idea of a hierarchy. In the current structure, you would have an Effect as a base, and all other effects are children of that. The structure is not complex, it would look like that:
```
| CardEffect
|
| ------ PoisonEffect
| ------ DamageEffect
| ------ HealEffect
| ------ DrawEffect
| ------ StatusEffect
| ...
```
I also think it would be easier to debug the effects one by one (ie file by file), than having a big file with everything in it. It's easier to identify where it doesn't work.
- I don't consider dealing damage, drawing cards and restoring AP "similar" effects at all. And here you can do all 3 things "at the same time", you just need to add the respective effects to the card.

The idea behind a JSON was to have a system where you could edit cards without having to touch the codebase at all. You want to create a new card ? Just make a file that list its effects and values, add it to a list of cards. Next time the game starts, it looks at the list of cards and creates them.

For the string parser it could be tricky yes. But it would mostly be a `to_str` for each class I feel like :thinking: 
It's definitely not a priority I think.

**Comment by cheesefrycook** [comment]
Thinking about it more, I think you’re right about the hierarchy. My worry was that we are going to have to rewrite code in lots of effects, but I suppose we can just write some helper functions that will do basic stuff like damage and buffing in one line. The hierarchy categories were meant to be basic player actions. Maybe that’s a helpful distinction for stuff later, maybe not. 

Resources will also not require you to touch the code base. It’s a separate file just like JSON. I consider Resources basically just a JSON file that interacts better with godot. It will throw errors for us as well if stuff is written wrong. The only downside I can think of is you have to open the godot editor to create the Resource initially. But once you do that you can edit it externally if you want.

**Comment by Turtyo** [comment]
Well we will need to rewrite code now. But I don't think we will to rewrite that many effects. There will only be as many effects as there are actions. I think the best is to not have "combined" effects (meaning effects that do two things at the same time, like damage and heal in the same effect). And if you want to make a card that is similar to an already existing card, you just copy the text file of said card (either the JSON or whatever format we decide to use) and you modify it.

You can probably create a resource file outside Godot as well. Nothing it set in stone for that part, so the discussion is open. The best would probably be the string parser for the display, so there are no discrepancies between what is said and what is actually done (though you could probably spot it easily). On the other hand, if the file has built-in error detection for functions and such, it's good too.

**Comment by cheesefrycook** [comment]
I think this will be a good one to implement after #17 is merged since this will affect a lot of stuff. I can work on this one unless you'd like to take it

**Comment by Turtyo** [comment]
I would like to take this one. I haven't touched code much and I have mostly done reviews and small PR. It would also help me if I had to work with the logic of everything here.

**Comment by Turtyo** [comment]
I think I will implement what was discussed at the end of #21 because it makes sense to do it now (cf [this comment](https://github.com/Saplings-Projects/1M_sub/issues/21#issuecomment-1858997970)

**Comment by Turtyo** [comment]
I will also need to do some reformating by using a list of entity for targets, because otherwise it's just not gonna work

**Comment by Turtyo** [comment]
Another thing that could be added now or later (though it would greatly simplify it if I do it now) is a targeting function. Here is the idea:
`EffectBase` has a basic targeting function, which adds to the list of enemy the one that is targeted by the player click. You can then override in child classes to target in a different way (for example, if you want to target the enemy clicked and adjacent enemies, or you want to target everyone)

This was not originally discussed in this issue so it might be better to have it in a new issue, but at the same time it would simplify my implementation of the card structure (which is precisely the goal of this issue)

**Comment by Tysterman74** [comment]
If you think that the targeting functioning is needed for this issue, then I'd say make the refactor now. We're at a somewhat early stage of development so making these needed refactors is better now rather than later once we have a bunch more files to modify. If the issue is amount of work needed to complete, I can help out with it!

**Comment by Turtyo** [comment]
I should be able to complete it like that thanks. I've also thought of a more advanced way to have targeting (by putting targeting functions into their own class instead of having them be functions of the effect, because you could want to damage one enemy, or damage multiple enemies, meaning you would need to create two effects for that when the only thing changing is who you are attacking).
But i'll keep that for a further issue to enhance the system. For now i just want to make it work and it can be changed later without too much trouble I think. I'll think about the current implementation before starting to make sure i'm not missing something. Will make more push in the PR tomorrow !
